README.md
# docker-ubuntu-desktop

docker-ubuntu-desktop 是一个 Docker 镜像，提供 Web VNC 接口来访问 Ubuntu Xfce4 桌面环境。

![desktop](./img/desktop.jpg)

![chrome](./img/chrome.jpg)

## 特性

- Ubuntu 20.04
- 桌面环境 [**Xfce4**](http://www.xfce.org)
- VNC服务 (默认 VNC 端口 `5901`)
- [**noVNC**](https://github.com/novnc/noVNC) - HTML5 VNC 客户端 (默认 http 端口 `6901`)
- 谷歌浏览器
- 拼音输入法

## 用法

- 打印帮助页:

```
docker run imgyh/docker-ubuntu-desktop --help
```

- 运行映射到本地端口 `5901`（vnc 协议）和 `6901`（vnc web 访问）的命令：

```
docker run -d -p 5901:5901 -p 6901:6901 imgyh/docker-ubuntu-desktop
```

- 进入容器使用交互模式 `-it` 和 `bash`

```
docker run -it -p 5901:5901 -p 6901:6901 imgyh/docker-ubuntu-desktop bash
```

## 连接与控制

如果容器像上面提到的那样启动，请通过以下选项之一进行连接：

- 通过 VNC 查看器 `localhost:5901` 连接，默认密码：`vncpassword`
- 通过 noVNC HTML5 完整客户端连接：[`http://localhost:6901`](http://localhost:6901)，默认密码：`vncpassword`
- 通过noVNC HTML5 lite 客户端连接：[`http://localhost:6901/vnc_lite.html?password=vncpassword`](http://localhost:6901/vnc_lite.html?password=vncpassword)

## 环境变量设置

可以在 `docker run` 阶段覆盖以下 VNC 环境变量以自定义容器内的桌面环境：

- `VNC_COL_DEPTH`， 默认：`24`
- `VNC_RESOLUTION`，默认：`1280x1024`
- `VNC_PW`，默认：`vncpassword`
- `VNC_PASSWORDLESS`，默认：`false`

### 重写VNC密码

覆盖环境变量的值 `VNC_PW`

```
docker run -it -p 5901:5901 -p 6901:6901 -e VNC_PW=my-pw imgyh/docker-ubuntu-desktop
```

### 重写 VNC 分辨率

覆盖环境变量的值 `VNC_RESOLUTION`

```
docker run -it -p 5901:5901 -p 6901:6901 -e VNC_RESOLUTION=800x600 imgyh/docker-ubuntu-desktop
```

### 无密码启动

设置 `VNC_PASSWORDLESS` 为 `true` 禁用 VNC 密码

```
docker run -it -p 5901:5901 -p 6901:6901 -e VNC_PASSWORDLESS=true imgyh/docker-ubuntu-desktop
```

### 仅查看 VNC

设置环境变量 `VNC_VIEW_ONLY=true`。如果设置，启动脚本将为控制连接创建一个随机密码保存在 `VNC_PW`环境变量中。可以连接到容器查看密码。
```
docker run -it -p 5901:5901 -p 6901:6901 -e VNC_VIEW_ONLY=true imgyh/docker-ubuntu-desktop
```

## 拼音输入法设置

找到Settings->iBus Preference->Input Method->Add->Chinese

![ibus1](./img/ibus1.jpg)

![ibus2](./img/ibus2.jpg)

![ibus3](./img/ibus3.jpg)

## 参考

[docker-ubuntu-desktop](https://github.com/gotoeasy/docker-ubuntu-desktop)

[docker-ubuntu-vnc-desktop](https://github.com/fcwu/docker-ubuntu-vnc-desktop)

[docker-headless-vnc-container](https://github.com/ConSol/docker-headless-vnc-container)

## 更新日志

更新日志在此处查看：[**changelog.md**](./changelog.md)

changelog.md
# docker-ubuntu-desktop 更新日志

## V0.1

- 安装一些工具：git curl vim wget net-tools locales bzip2 procps
- 安装 Xfce4 桌面环境
- 安装 tigervnc
- 安装 no_vnc
- 安装 chrome 浏览器
- 安装 ibus拼音输入法

src/scripts/vnc_startup.sh
#!/usr/bin/env bash
### every exit != 0 fails the script
set -e

## print out help
help (){
echo "
USAGE:
docker run -itd -p 6901:6901 -p 5901:5901 imgyh/<image>:<tag> <option>

IMAGES:
	docker-ubuntu-desktop

TAGS:
latest  stable version of branch 'main'

OPTIONS:
-w, --wait      (default) keeps the UI and the vncserver up until SIGINT or SIGTERM will received
-s, --skip      skip the vnc startup and just execute the assigned command.
                example: docker run imgyh/docker-ubuntu-desktop --skip bash
-d, --debug     enables more detailed startup output
                e.g. 'docker run imgyh/docker-ubuntu-desktop --debug bash'
-h, --help      print out this help

Fore more information see: https://github.com/imgyh/docker-ubuntu-desktop
"
}
if [[ $1 =~ -h|--help ]]; then
    help
    exit 0
fi


# add `--skip` to startup args, to skip the VNC startup procedure
if [[ $1 =~ -s|--skip ]]; then
    echo -e "\n\n------------------ SKIP VNC STARTUP -----------------"
    echo -e "\n\n------------------ EXECUTE COMMAND ------------------"
    echo "Executing command: '${@:2}'"
    exec "${@:2}"
fi
if [[ $1 =~ -d|--debug ]]; then
    echo -e "\n\n------------------ DEBUG VNC STARTUP -----------------"
    export DEBUG=true
fi

## correct forwarding of shutdown signal
cleanup () {
    kill -s SIGTERM $!
    exit 0
}
trap cleanup SIGINT SIGTERM

## write correct window size to chrome properties
$STARTUPDIR/chrome-init.sh

## resolve_vnc_connection
VNC_IP=$(hostname -i)

## change vnc password
echo -e "\n------------------ change VNC password  ------------------"
# first entry is control, second is view (if only one is valid for both)
mkdir -p "$HOME/.vnc"
PASSWD_PATH="$HOME/.vnc/passwd"

if [[ -f $PASSWD_PATH ]]; then
    echo -e "\n---------  purging existing VNC password settings  ---------"
    rm -f $PASSWD_PATH
fi

if [[ $VNC_VIEW_ONLY == "true" ]]; then
    echo "start VNC server in VIEW ONLY mode!"
    #create random pw to prevent access
    echo $(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 20) | vncpasswd -f > $PASSWD_PATH
fi
echo "$VNC_PW" | vncpasswd -f >> $PASSWD_PATH
chmod 600 $PASSWD_PATH


## start vncserver and noVNC webclient
echo -e "\n------------------ start noVNC  ----------------------------"
if [[ $DEBUG == true ]]; then echo "$NO_VNC_HOME/utils/novnc_proxy --vnc localhost:$VNC_PORT --listen $NO_VNC_PORT"; fi
$NO_VNC_HOME/utils/novnc_proxy --vnc localhost:$VNC_PORT --listen $NO_VNC_PORT > $STARTUPDIR/no_vnc_startup.log 2>&1 &
PID_SUB=$!

echo -e "start window manager\n..."
$STARTUPDIR/wm_startup.sh &> $STARTUPDIR/wm_startup.log


#echo -e "\n------------------ start VNC server ------------------------"
#echo "remove old vnc locks to be a reattachable container"
vncserver -kill $DISPLAY &> $STARTUPDIR/vnc_startup.log \
    || rm -rfv /tmp/.X*-lock /tmp/.X11-unix &> $STARTUPDIR/vnc_startup.log \
    || echo "no locks present"

echo -e "start vncserver with param: VNC_COL_DEPTH=$VNC_COL_DEPTH, VNC_RESOLUTION=$VNC_RESOLUTION\n..."

vnc_cmd="vncserver $DISPLAY -depth $VNC_COL_DEPTH -geometry $VNC_RESOLUTION PasswordFile=$HOME/.vnc/passwd"
if [[ ${VNC_PASSWORDLESS:-} == "true" ]]; then
  vnc_cmd="${vnc_cmd} -SecurityTypes None"
fi

if [[ $DEBUG == true ]]; then echo "$vnc_cmd"; fi
$vnc_cmd > $STARTUPDIR/no_vnc_startup.log 2>&1

### ibus-pinyin start
export GTK_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=ibus
ibus-daemon -rdx

## log connect options
echo -e "\n\n------------------ VNC environment started ------------------"
echo -e "\nVNCSERVER started on DISPLAY= $DISPLAY \n\t=> connect via VNC viewer with $VNC_IP:$VNC_PORT"
echo -e "\nnoVNC HTML client started:\n\t=> connect via http://$VNC_IP:$NO_VNC_PORT/?password=...\n"


if [[ $DEBUG == true ]] || [[ $1 =~ -t|--tail-log ]]; then
    echo -e "\n------------------ $HOME/.vnc/*$DISPLAY.log ------------------"
    # if option `-t` or `--tail-log` block the execution and tail the VNC log
    tail -f $STARTUPDIR/*.log $HOME/.vnc/*$DISPLAY.log
fi

if [ -z "$1" ] || [[ $1 =~ -w|--wait ]]; then
    wait $PID_SUB
else
    # unknown option ==> call command
    echo -e "\n\n------------------ EXECUTE COMMAND ------------------"
    echo "Executing command: '$@'"
    exec "$@"
fi
